### PR TITLE
[Developer Documentation] Illustrate the Agent upgrade process

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR adds developer documentation illustrating the Fleet-managed Agent upgrade process, focussing on the communications amongst the various components involved.

## Why is it important?

For Agent and Fleet developers to understand how Agent upgrades work.

## Review suggestions

View the rich diff of the file.

### CODE REVIEW

**Review by jlind23 — APPROVED** [review]

File: docs/upgrades.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+       UW->>UM: Remove
+       FS->>ES: Update Agent doc in `.fleet-agents`<br />set `upgrade_status` = null<br />`upgraded_at = <now>
+       UI->>UI: Show Agent status as "healthy"
+   end
+   opt Rollback
+       UW->>A: Start
+       A->>FS: Ack failed upgrade
+       FS->>ES: Update Agent doc in `.fleet-agents`<br />set `upgrade_status` = null<br />`upgraded_at = <now>
+       Note right of ES: Need to check
+       UI->>UI: Show Agent status as "???"
```
Comment (ycombinator) [inline docs/upgrades.md:54]: FIXME: replace `???`.

File: docs/upgrades.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+       Note right of UI: Need to check
```
Comment (ycombinator) [inline docs/upgrades.md:55]: TODO: check and remove note.

File: docs/upgrades.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    UI->>ES: Create new doc in `.fleet-actions` for `UPGRADE` action
+    A->>FS: Check-in request
+    FS->>ES: Read pending actions from .fleet-actions
+    FS->>A: Check-in response
+    A->>A: Queue upgrade action
+    alt If upgrade start fails
+       A->>FS: Ack failed upgrade
+       FS->>ES: Update Agent doc in `.fleet-agents`<br />set `upgrade_status` = "failed"
+       UI->>UI: Show Agent status as "???"
+       Note right of UI: Need to check
```
Comment (ycombinator) [inline docs/upgrades.md:32]: TODO: check and remove note.

Comment (juliaElastic) [reply docs/upgrades.md:32]: Yeah I think the agent status stays in updating.  I think in some cases the error is written to `.fleet-action-results` and the UI shows the action failure in Agent activity flyout.

File: docs/upgrades.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+       end
+       A->>A: Download new Agent artifact
+       A->>A: Extract new Agent artifact
+       A->>A: Replace current Agent artifact with new one
+       A->>UM: Create
+       A->>UW: Start
+       A->>A: Rexec to start new Agent artifact
+       A->>FS: Ack successful upgrade
+       UW->>UM: Remove
+       FS->>ES: Update Agent doc in `.fleet-agents`<br />set `upgrade_status` = null<br />`upgraded_at = <now>
```
Comment (juliaElastic) [inline docs/upgrades.md:46]: also set `upgrade_started_at` = null

Comment (juliaElastic) [reply docs/upgrades.md:46]: There is also a step here where FS writes the successful ack to `.fleet-actions-results`. We have seen an issue where the write fails (e.g. ES connection error) and the action stays in progress in the UI forever.

File: docs/upgrades.md (hunk 2 above)
Comment (juliaElastic) [reply docs/upgrades.md:55]: Agent status goes back to Healthy on successful rollback (by clearing `upgrade_started_at` field)

**Review by juliaElastic — APPROVED** [review]

**Pull request merged**